File: leetcode/gray_code.py
# Given a non-negative integer n representing the total number of bits in 
# the code, print the sequence of gray code. 
# A gray code sequence must begin with 0.
# Time Complexity: O(2^n)
# Space Complexity: O(2^n)

class Solution:
    # @param {integer} n
    # @return {integer[]}
    def grayCode(self, n):
        return [(x >> 1) ^ x for x in xrange(2 ** n)]

if __name__ == "__main__":
    solution = Solution()
    print solution.grayCode(2)

File: leetcode/gray_code.js
/* Given a non-negative integer n representing the total number of bits in 
 * the code, print the sequence of gray code. 
 * A gray code sequence must begin with 0.
 * Time Complexity: O(2^n)
 * Space Complexity: O(2^n)
 */

/**
 * @param {number} n
 * @return {number[]}
 */
var grayCode = function(n) {
    var result = []
    for (var x = 0; x < Math.pow(2, n); ++x)
        result.push((x >> 1) ^ x)
    return result
};
